src/typecheck/mod.rs

```diff
@@ -339,65 +385,75 @@ impl UnifError {
     }
 }
 
+/// Populate a new global typing environment from a global term environment.
+pub fn mk_global(eval_env: &eval::Environment) -> GlobalEnvironment {
+    eval_env
+        .iter()
+        .map(|(id, thunk)| {
+            (
+                id.clone(),
+                apparent_type(thunk.borrow().body.as_ref(), None).into(),
+            )
+        })
+        .collect()
+}
+
 /// The typing environment.
-pub type Environment = GenericEnvironment<Ident, TypeWrapper>;
+pub type GlobalEnvironment = GenericEnvironment<Ident, TypeWrapper>;
+pub type TypeEnvironment = GenericEnvironment<Ident, Closure>;
 
+impl From<&GlobalEnvironment> for TypeEnvironment {
+    fn from(g: &GlobalEnvironment) -> Self {
+        let mut env = TypeEnvironment::new();
+        for (id, tw) in g.iter() {
+            env.insert(id.clone(), tw.into());
+        }
```

Comment:
I don't understand what you mean, the `&GlobalEnvironment` argument is still passed in the `type_check` (and other) functions for this purpose

Reply:
Note that right now we are not dealing with type aliases yet, even if this PR is made for this feature request.

I don't understand the difference between the local typing environment and the local typealias environment, and why it needs to be 2 separated environments when we could merge the two of them and use them inside Closures.

The reason I didn't merged them is that the local "typealias" environment is made for explicitly defined type bindings (and to add a simple layer instead of breaking the existing system)